File: src/web/Controllers/AlertsController.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using core.fs.Adapters.Logging;
using core.fs.Alerts;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

#nullable enable

namespace web.Controllers;

[ApiController]
[Authorize]
[Route("api/[controller]")]
public class AlertsController(Handler handler) : ControllerBase
{
    // TODO: this is exposed anonymously to allow trendview to ping this endpoint
    // and they don't support authorization. I should at the very least set up some sort of query
    // string key or something along those lines
    [AllowAnonymous]
    [Consumes("text/plain")]
    [HttpPost("sms")]
    public Task Sms([FromBody] string body) =>
        handler.Handle(new SendSMS(body));

    [AllowAnonymous]
    [HttpGet("sms/status")]
    public Task<ActionResult> SmsStatus() =>
        this.OkOrError(handler.Handle(new SMSStatus()));

    [Authorize("admin")]
    [HttpPost("sms/on")]
    public Task SmsOn() => handler.Handle(new TurnSMSOn());

    [Authorize("admin")]
    [HttpPost("sms/off")]
    public Task SmsOff() => handler.Handle(new TurnSMSOff());

    [Authorize("admin")]
    [HttpPost("run")]
    public ActionResult Run() => this.OkOrError(handler.Handle(new Run()));

    [HttpGet]
    public AlertsView Index() => handler.Handle(new QueryAlerts(User.Identifier()));

    [HttpGet("monitors")]
    public IEnumerable<object> Monitors() => handler.Handle(new QueryAvailableMonitors());
        
    [HttpGet("emailJob")]
    [Authorize("admin")]
    public Task EmailJob() => handler.Handle(new RunEmailJob());

    [HttpGet("triggerWeekly")]
    [Authorize("admin")]
    public Task TriggerWeekly(
        [FromServices] ILogger logger,
        [FromServices] MonitoringServices.WeeklyMonitoringService weeklyAlerts) => weeklyAlerts.Execute(true);

    [HttpGet("triggerMaxProfit")]
    [Authorize("admin")]
    public Task TriggerMaxProfit(
        [FromServices] core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService service) => service.ReportOnMaxProfitBasedOnDaysHeld();
    
    [HttpGet("triggerThirtyDayTransactions")]
    [Authorize("admin")]
    public Task TriggerThirtyDayTransactions(
        [FromServices] core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService service) => service.ReportOnThirtyDayTransactions();
        
    [HttpGet("triggerDaily")]
    [Authorize("admin")]
    public Task TriggerDaily([FromServices] MonitoringServices.PatternMonitoringService dailyAlerts)
    {
        dailyAlerts.RunPatternMonitoring();
        return Task.CompletedTask;
    }

    [HttpGet("triggerPriceObvTrend")]
    [Authorize("admin")]
    public Task TriggerPriceObvTrend([FromServices] MonitoringServices.PriceObvTrendMonitoringService priceObvTrendAlerts)
    {
        priceObvTrendAlerts.Run();
        return Task.CompletedTask;
    }
        
    [HttpGet("triggerEmail")]
    [Authorize("admin")]
    public Task TriggerEmail(
        [FromServices] MonitoringServices.AlertEmailService emailAlerts)
    {
        emailAlerts.Execute();
        return Task.CompletedTask;
    }
        
    [HttpGet("triggerAccountMonitoring")]
    [Authorize("admin")]
    public Task TriggerAccountMonitoring(
        [FromServices] core.fs.Brokerage.MonitoringServices.AccountMonitoringService accountMonitoring)
        => accountMonitoring.RunAccountValueOrderAndTransactionSync();
    
    [HttpGet("triggerAccountTransactionProcessing")]
    [Authorize("admin")]
    public Task TriggerAccountTransactionProcessing(
        [FromServices] core.fs.Brokerage.MonitoringServices.AccountMonitoringService accountMonitoring)
        => accountMonitoring.RunTransactionProcessing();
    
    [HttpGet("triggerPortfolioProcessing")]
    [Authorize("admin")]
    public Task TriggerPortfolioProcessing(
        [FromServices] core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService service)
        => service.RecentlyClosedPositionUpdates();
    
    [HttpGet("triggerOptionPricing")]
    [Authorize("admin")]
    public Task TriggerOptionPricing(
        [FromServices] core.fs.Options.MonitoringServices.PriceMonitoringService optionPricing)
        => optionPricing.Run();
    
    [HttpGet("triggerPriceAlerts")]
    [Authorize("admin")]
    public Task TriggerPriceAlerts(
        [FromServices] MonitoringServices.PriceAlertMonitoringService priceAlerts)
        => priceAlerts.Execute();
    
    [HttpGet("triggerNearTriggerPriceAlerts")]
    [Authorize("admin")]
    public Task TriggerNearTriggerPriceAlerts(
        [FromServices] MonitoringServices.PriceAlertNearTriggerMonitoringService nearTriggerAlerts)
        => nearTriggerAlerts.Execute();
    
    [HttpGet("triggerReminders")]
    [Authorize("admin")]
    public Task TriggerReminders(
        [FromServices] MonitoringServices.ReminderMonitoringService reminderService)
        => reminderService.Execute();
    
    [HttpGet("triggerSECFilings")]
    [Authorize("admin")]
    public Task TriggerSECFilings(
        [FromServices] SECFilingsMonitoring.SECFilingsMonitoringService secFilingsService)
        => secFilingsService.Execute();
    
    // Stock Price Alerts CRUD
    [HttpGet("price")]
    public Task<ActionResult> GetStockPriceAlerts() =>
        this.OkOrError(handler.Handle(new GetStockPriceAlerts(User.Identifier())));
    
    [HttpPost("price")]
    public Task<ActionResult> CreateStockPriceAlert([FromBody] CreateStockPriceAlertRequest request) =>
        this.OkOrError(handler.Handle(new CreateStockPriceAlert(
            userId: User.Identifier(),
            ticker: request.Ticker,
            priceLevel: request.PriceLevel,
            alertType: request.AlertType,
            note: request.Note
        )));
    
    [HttpPut("price/{alertId}")]
    public Task<ActionResult> UpdateStockPriceAlert([FromRoute] System.Guid alertId, [FromBody] UpdateStockPriceAlertRequest request) =>
        this.OkOrError(handler.Handle(new UpdateStockPriceAlert(
            userId: User.Identifier(),
            alertId: alertId,
            priceLevel: request.PriceLevel,
            alertType: request.AlertType,
            note: request.Note,
            state: request.State
        )));
    
    [HttpPost("price/{alertId}/reset")]
    public Task<ActionResult> ResetStockPriceAlert([FromRoute] System.Guid alertId) =>
        this.OkOrError(handler.Handle(new ResetStockPriceAlert(
            userId: User.Identifier(),
            alertId: alertId
        )));
    
    [HttpDelete("price/{alertId}")]
    public Task<ActionResult> DeleteStockPriceAlert([FromRoute] System.Guid alertId) =>
        this.OkOrError(handler.Handle(new DeleteStockPriceAlert(alertId)));
    
    // Reminders CRUD
    [HttpGet("reminders")]
    public Task<ActionResult> GetReminders() =>
        this.OkOrError(handler.Handle(new GetReminders(User.Identifier())));
    
    [HttpPost("reminders")]
    public Task<ActionResult> CreateReminder([FromBody] CreateReminderRequest request) =>
        this.OkOrError(handler.Handle(new CreateReminder(
            userId: User.Identifier(),
            date: request.Date,
            message: request.Message,
            ticker: request.Ticker
        )));
    
    [HttpPut("reminders/{reminderId}")]
    public Task<ActionResult> UpdateReminder([FromRoute] System.Guid reminderId, [FromBody] UpdateReminderRequest request) =>
        this.OkOrError(handler.Handle(new UpdateReminder(
            userId: User.Identifier(),
            reminderId: reminderId,
            date: request.Date,
            message: request.Message,
            ticker: request.Ticker,
            state: request.State
        )));
    
    [HttpDelete("reminders/{reminderId}")]
    public Task<ActionResult> DeleteReminder([FromRoute] System.Guid reminderId) =>
        this.OkOrError(handler.Handle(new DeleteReminder(reminderId)));
}

// Request DTOs for API
public record CreateStockPriceAlertRequest(string Ticker, decimal PriceLevel, string AlertType, string Note);
public record UpdateStockPriceAlertRequest(decimal PriceLevel, string AlertType, string Note, string State);
public record CreateReminderRequest(string Date, string Message, string? Ticker);
public record UpdateReminderRequest(string Date, string Message, string? Ticker, string State);

File: src/web.fs/Controllers/AlertsController.fs
namespace web.Controllers

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Adapters.Logging
open core.fs.Alerts
open web.Utils

[<CLIMutable>]
type CreateStockPriceAlertRequest = {
    Ticker: string
    PriceLevel: decimal
    AlertType: string
    Note: string
}

[<CLIMutable>]
type UpdateStockPriceAlertRequest = {
    PriceLevel: decimal
    AlertType: string
    Note: string
    State: string
}

[<CLIMutable>]
type CreateReminderRequest = {
    Date: string
    Message: string
    Ticker: string option
}

[<CLIMutable>]
type UpdateReminderRequest = {
    Date: string
    Message: string
    Ticker: string option
    State: string
}

[<ApiController>]
[<Authorize>]
[<Route("api/[controller]")>]
type AlertsController(handler: Handler) =
    inherit ControllerBase()

    [<AllowAnonymous>]
    [<Consumes("text/plain")>]
    [<HttpPost("sms")>]
    member this.Sms([<FromBody>] body: string) =
        handler.Handle({SendSMS.Body = body})

    [<AllowAnonymous>]
    [<HttpGet("sms/status")>]
    member this.SmsStatus() =
        this.OkOrError(handler.Handle(SMSStatus()))

    [<Authorize("admin")>]
    [<HttpPost("sms/on")>]
    member this.SmsOn() =
        handler.Handle(TurnSMSOn())

    [<Authorize("admin")>]
    [<HttpPost("sms/off")>]
    member this.SmsOff() =
        handler.Handle(TurnSMSOff())

    [<Authorize("admin")>]
    [<HttpPost("run")>]
    member this.Run() =
        this.OkOrError(handler.Handle(Run()))

    [<HttpGet>]
    member this.Index() =
        handler.Handle({QueryAlerts.UserId = this.User.Identifier()})

    [<HttpGet("monitors")>]
    member this.Monitors() =
        handler.Handle(QueryAvailableMonitors())

    [<HttpGet("emailJob")>]
    [<Authorize("admin")>]
    member this.EmailJob() =
        handler.Handle(RunEmailJob())

    [<HttpGet("triggerWeekly")>]
    [<Authorize("admin")>]
    member this.TriggerWeekly([<FromServices>] logger: ILogger, [<FromServices>] weeklyAlerts: MonitoringServices.WeeklyMonitoringService) =
        weeklyAlerts.Execute(true)

    [<HttpGet("triggerMaxProfit")>]
    [<Authorize("admin")>]
    member this.TriggerMaxProfit([<FromServices>] service: core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService) =
        service.ReportOnMaxProfitBasedOnDaysHeld()

    [<HttpGet("triggerThirtyDayTransactions")>]
    [<Authorize("admin")>]
    member this.TriggerThirtyDayTransactions([<FromServices>] service: core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService) =
        service.ReportOnThirtyDayTransactions()

    [<HttpGet("triggerDaily")>]
    [<Authorize("admin")>]
    member this.TriggerDaily([<FromServices>] dailyAlerts: MonitoringServices.PatternMonitoringService) = task {
        return! dailyAlerts.RunPatternMonitoring()
    }

    [<HttpGet("triggerPriceObvTrend")>]
    [<Authorize("admin")>]
    member this.TriggerPriceObvTrend([<FromServices>] priceObvTrendAlerts: MonitoringServices.PriceObvTrendMonitoringService) = task {
        return! priceObvTrendAlerts.Run()
    }

    [<HttpGet("triggerEmail")>]
    [<Authorize("admin")>]
    member this.TriggerEmail([<FromServices>] emailAlerts: MonitoringServices.AlertEmailService) = task {
        return! emailAlerts.Execute()
    }

    [<HttpGet("triggerAccountMonitoring")>]
    [<Authorize("admin")>]
    member this.TriggerAccountMonitoring([<FromServices>] accountMonitoring: core.fs.Brokerage.MonitoringServices.AccountMonitoringService) =
        accountMonitoring.RunAccountValueOrderAndTransactionSync()

    [<HttpGet("triggerAccountTransactionProcessing")>]
    [<Authorize("admin")>]
    member this.TriggerAccountTransactionProcessing([<FromServices>] accountMonitoring: core.fs.Brokerage.MonitoringServices.AccountMonitoringService) =
        accountMonitoring.RunTransactionProcessing()

    [<HttpGet("triggerPortfolioProcessing")>]
    [<Authorize("admin")>]
    member this.TriggerPortfolioProcessing([<FromServices>] service: core.fs.Portfolio.MonitoringServices.PortfolioAnalysisService) =
        service.RecentlyClosedPositionUpdates()

    [<HttpGet("triggerOptionPricing")>]
    [<Authorize("admin")>]
    member this.TriggerOptionPricing([<FromServices>] optionPricing: core.fs.Options.MonitoringServices.PriceMonitoringService) =
        optionPricing.Run()

    [<HttpGet("triggerPriceAlerts")>]
    [<Authorize("admin")>]
    member this.TriggerPriceAlerts([<FromServices>] priceAlerts: MonitoringServices.PriceAlertMonitoringService) =
        priceAlerts.Execute()

    [<HttpGet("triggerNearTriggerPriceAlerts")>]
    [<Authorize("admin")>]
    member this.TriggerNearTriggerPriceAlerts([<FromServices>] nearTriggerAlerts: MonitoringServices.PriceAlertNearTriggerMonitoringService) =
        nearTriggerAlerts.Execute()

    [<HttpGet("triggerReminders")>]
    [<Authorize("admin")>]
    member this.TriggerReminders([<FromServices>] reminderService: MonitoringServices.ReminderMonitoringService) =
        reminderService.Execute()

    [<HttpGet("triggerSECFilingsSync")>]
    [<Authorize("admin")>]
    member this.TriggerSECFilingsSync([<FromServices>] secFilingsSyncService: SECFilingsMonitoring.SECFilingsSyncService) =
        secFilingsSyncService.Execute()

    [<HttpGet("triggerSECFilings")>]
    [<Authorize("admin")>]
    member this.TriggerSECFilings([<FromServices>] secFilingsService: SECFilingsMonitoring.SECFilingsMonitoringService) =
        secFilingsService.Execute()

    [<HttpGet("triggerSchedule13G")>]
    [<Authorize("admin")>]
    member this.TriggerSchedule13G([<FromServices>] service: secedgar.fs.Schedule13GProcessingService) =
        service.Execute()

    [<HttpGet("triggerForm144")>]
    [<Authorize("admin")>]
    member this.TriggerForm144([<FromServices>] service: secedgar.fs.Form144ProcessingService) =
        service.Execute()

    [<HttpGet("triggerSchedule13D")>]
    [<Authorize("admin")>]
    member this.TriggerSchedule13D([<FromServices>] service: secedgar.fs.Schedule13DProcessingService) =
        service.Execute()

    [<HttpGet("price")>]
    member this.GetStockPriceAlerts() =
        this.OkOrError(handler.Handle({GetStockPriceAlerts.UserId = this.User.Identifier()}))

    [<HttpPost("price")>]
    member this.CreateStockPriceAlert([<FromBody>] request: CreateStockPriceAlertRequest) =
        this.OkOrError(handler.Handle({
            CreateStockPriceAlert.UserId = this.User.Identifier()
            Ticker = request.Ticker
            PriceLevel = request.PriceLevel
            AlertType = request.AlertType
            Note = request.Note
        }))

    [<HttpPut("price/{alertId}")>]
    member this.UpdateStockPriceAlert([<FromRoute>] alertId: Guid, [<FromBody>] request: UpdateStockPriceAlertRequest) =
        this.OkOrError(handler.Handle({
            UpdateStockPriceAlert.UserId = this.User.Identifier()
            AlertId = alertId
            PriceLevel = request.PriceLevel
            AlertType = request.AlertType
            Note = request.Note
            State = request.State
        }))

    [<HttpPost("price/{alertId}/reset")>]
    member this.ResetStockPriceAlert([<FromRoute>] alertId: Guid) =
        this.OkOrError(handler.Handle({
            ResetStockPriceAlert.UserId = this.User.Identifier()
            AlertId = alertId
        }))

    [<HttpDelete("price/{alertId}")>]
    member this.DeleteStockPriceAlert([<FromRoute>] alertId: Guid) =
        this.OkOrError(handler.Handle({DeleteStockPriceAlert.UserId = this.User.Identifier(); AlertId = alertId}))

    [<HttpGet("reminders")>]
    member this.GetReminders() =
        this.OkOrError(handler.Handle({GetReminders.UserId = this.User.Identifier()}))

    [<HttpPost("reminders")>]
    member this.CreateReminder([<FromBody>] request: CreateReminderRequest) =
        this.OkOrError(handler.Handle({
            CreateReminder.UserId = this.User.Identifier()
            Date = request.Date
            Message = request.Message
            Ticker = request.Ticker
        }))

    [<HttpPut("reminders/{reminderId}")>]
    member this.UpdateReminder([<FromRoute>] reminderId: Guid, [<FromBody>] request: UpdateReminderRequest) =
        this.OkOrError(handler.Handle({
            UpdateReminder.UserId = this.User.Identifier()
            ReminderId = reminderId
            Date = request.Date
            Message = request.Message
            Ticker = request.Ticker
            State = request.State
        }))

    [<HttpDelete("reminders/{reminderId}")>]
    member this.DeleteReminder([<FromRoute>] reminderId: Guid) =
        this.OkOrError(handler.Handle({DeleteReminder.UserId = this.User.Identifier(); ReminderId = reminderId}))
